From c36d4af8a88190c745f1308f356398fb46f1934d Mon Sep 17 00:00:00 2001
From: Zaikuo Wang <zaikuo.wang@intel.com>
Date: Fri, 20 Mar 2015 11:23:23 +0800
Subject: [PATCH 312/314] atomisp: fix panic in atomisp_css_recover

IPANIC on ia_css_pipe_get_pipeline, because ACC pipe pointer
is NULL

Change-Id: Ic5eb8144e9d5c04ee85d01ef18079a2fb20b1195
Tracked-On: #JIMINAN-26386
Signed-off-by: Zaikuo Wang <zaikuo.wang@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/47193
Reviewed-by: Jian Xu Zheng <jian.xu.zheng@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Approver: Jian Xu Zheng <jian.xu.zheng@intel.com>
Integrator: Wei Tao <wei.tao@intel.com>
Build: IMIN-legacy (SC)
Reviewed-by: Tian Shu Qiu <tian.shu.qiu@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp_cmd.c      | 22 ++++++++++++----------
 1 file changed, 12 insertions(+), 10 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index 7d37077..ac3477b 100755
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -1246,6 +1246,7 @@ static void __atomisp_css_recover(struct atomisp_device *isp, bool isp_timeout)
 	for (i = 0; i < isp->num_of_streams; i++) {
 		struct atomisp_sub_device *asd = &isp->asd[i];
 		struct ia_css_pipeline *acc_pipeline;
+		struct ia_css_pipe *acc_pipe = NULL;
 
 		if (asd->streaming != ATOMISP_DEVICE_STREAMING_ENABLED &&
 		    !asd->stream_prepared)
@@ -1256,16 +1257,17 @@ static void __atomisp_css_recover(struct atomisp_device *isp, bool isp_timeout)
 		* By calling acc_done() for all loaded fw_handles,
 		* HAL will be unblocked.
 		*/
-		acc_pipeline =
-			ia_css_pipe_get_pipeline(
-			asd->stream_env[i].pipes[CSS_PIPE_ID_ACC]);
-		if (acc_pipeline) {
-			struct ia_css_pipeline_stage *stage;
-			for (stage = acc_pipeline->stages; stage;
-				stage = stage->next) {
-				const struct ia_css_fw_info *fw;
-				fw = stage->firmware;
-				atomisp_acc_done(asd, fw->handle);
+		acc_pipe = asd->stream_env[i].pipes[CSS_PIPE_ID_ACC];
+		if (acc_pipe != NULL) {
+			acc_pipeline = ia_css_pipe_get_pipeline(acc_pipe);
+			if (acc_pipeline) {
+				struct ia_css_pipeline_stage *stage;
+				for (stage = acc_pipeline->stages; stage;
+					stage = stage->next) {
+					const struct ia_css_fw_info *fw;
+					fw = stage->firmware;
+					atomisp_acc_done(asd, fw->handle);
+				}
 			}
 		}
 
-- 
1.9.1

